crates/turborepo-cache/src/http.rs

```diff
@@ -0,0 +1,253 @@
+    ) -> Result<(), CacheError> {
+        let mut cache_archive = CacheWriter::from_writer(writer, true)?;
+        for file in files {
+            cache_archive.add_file(anchor, &file)?;
+        }
+
+        Ok(())
+    }
+
+    pub async fn exists(
+        &self,
+        hash: &str,
+        token: &str,
+        team_id: &str,
+        team_slug: Option<&str>,
+        use_preflight: bool,
+    ) -> Result<u32, CacheError> {
+        let response = self
+            .client
+            .artifact_exists(hash, token, team_id, team_slug, use_preflight)
+            .await?;
+
+        let duration = Self::get_duration_from_response(&response)?;
+
+        Ok(duration)
+    }
+
+    fn get_duration_from_response(response: &Response) -> Result<u32, CacheError> {
+        if let Some(duration) = response.headers().get("x-artifact-duration") {
+            let duration = duration
+                .to_str()
+                .map_err(|_| CacheError::InvalidDuration(Backtrace::capture()))?;
+
+            duration
+                .parse::<u32>()
+                .map_err(|_| CacheError::InvalidDuration(Backtrace::capture()))
+        } else {
+            Ok(0)
+        }
+    }
+
+    pub async fn retrieve(
+        &self,
+        hash: &str,
+        token: &str,
+        team_id: &str,
+        team_slug: Option<&str>,
+        use_preflight: bool,
+    ) -> Result<(Vec<AnchoredSystemPathBuf>, u32), CacheError> {
+        let response = self
+            .client
+            .fetch_artifact(hash, token, team_id, team_slug, use_preflight)
+            .await?;
+
+        let duration = Self::get_duration_from_response(&response)?;
+
+        let body = if let Some(signer_verifier) = &self.signer_verifier {
+            let expected_tag = response
+                .headers()
+                .get("x-artifact-tag")
+                .ok_or(CacheError::ArtifactTagMissing(Backtrace::capture()))?;
+
+            let expected_tag = expected_tag
+                .to_str()
+                .map_err(|_| CacheError::InvalidTag(Backtrace::capture()))?
+                .to_string();
+
+            let body = response.bytes().await.map_err(|e| {
+                CacheError::ApiClientError(
+                    turborepo_api_client::Error::ReqwestError(e),
+                    Backtrace::capture(),
+                )
+            })?;
+            let is_valid = signer_verifier.validate(hash.as_bytes(), &body, &expected_tag)?;
+
+            if !is_valid {
+                return Err(CacheError::InvalidTag(Backtrace::capture()));
+            }
+
+            body
+        } else {
+            response.bytes().await.map_err(|e| {
+                CacheError::ApiClientError(
+                    turborepo_api_client::Error::ReqwestError(e),
+                    Backtrace::capture(),
+                )
+            })?
+        };
+
+        let files = Self::restore_tar(&self.repo_root, &body)?;
+
+        Ok((files, duration))
```

Comment:
Can we pretty please not return a tuple here?

Reply:
Would you prefer a struct? IMO a tuple is just fine here since they are separate types

Reply:
I basically always prefer structs because on numerous occasions I have screwed up tuples in a way where having the name would have rescued me. Yeah, they're different types now, but code changes over time and we end up with the ridiculous return values we have in the Go version.

Reply:
Actually, this feedback comment doesn't follow my rule regarding code style linting (enforced automatically with a linter or GTFO), Clippy doesn't have this lint to enable, so I'm marking this as resolved.

I still hold this opinion, but it's not blocking unless I make it automated.